File: 02_algorithms/01_number_algorithms/01_prime_number/prime_number.js
/**
 * Complexity
 *
 * primeNumberWay01 -> O(n)
 * primeNumberWay02 -> O(sqrt(n))
 */

function printPrime(number) { console.log(`${number} is Prime`) }

function primeNumberWay01(number) {
  if (number <= 1) {
    return
  }

  for (let i = 2; i < number; i++) {
    if (number % i == 0) {
      return
    }
  }

  return printPrime(number)
}

/**
 * @name primeNumberWay02
 * @description 6k + 1, to k > 0, excluding {2, 3, 5}
 * @param {number} int
 *
 * @returns {undefined, but print in console}
 */
function primeNumberWay02(number) {

  if (number <= 1)
    return;
  if (number <= 3)
    return printPrime(number);

  // skip middle five number between [i, i+6]
  if (number % 2 == 0 || number % 3 == 0)
    return;

  /**
   * test until square root of number
   * 'cause n*n != prime -> n != prime
   */
  for (let i = 5; i * i <= number; i = i + 6) {
    if (number % i == 0 || number % (i + 2) == 0) {
      return
    }
  }

  return printPrime(number)
}

function main() {

  for (let x = 0; x <= 10e4; x++) {
    // primeNumberWay01(x)
    primeNumberWay02(x)
  }
}

main()

File: 02_algorithms/01_number_algorithms/01_prime_number/prime_number.py
"""
Complexity

prime_number_way_01 -> O(n)
prime_number_way_02 -> O(sqrt(n))
"""


def print_prime(number: int):
    print(f'{number} is Prime')


def prime_number_way_01(number: int) -> None:
    if number <= 1:
        return

    for i in range(2, number, 1):
        if number % i == 0:
            return

    print_prime(number)


def prime_number_way_02(number: int) -> None:

    if number <= 1:
        return

    if number <= 3:
        print_prime(number)

    # skip middle five number between [i, i+6]
    if number % 2 == 0 or number % 3 == 0:
        return

    """
    test until square root of number
    'cause n*n != prime -> n != prime
    """
    i = 5
    for i in range(5, i * i <= number, 6):
        if number % i == 0 or number % (i + 2) == 0:
            return

    print_prime(number)


if __name__ == '__main__':
    for x in range(0, int(10e4), 1):
        # prime_number_way_01(x)
        prime_number_way_02(x)
